flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java

```diff
@@ -0,0 +1,272 @@
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a int", "c1 varchar", "c2 varchar")
+                                        .consumedBeforeRestore(
+                                                Row.of(1, "HiHi", "Hi2"),
+                                                Row.of(1, "HiHi", "Hi2"),
+                                                Row.of(1, "HiHi", "Hi3"),
+                                                Row.of(2, "HeHe", "Hi5"),
+                                                Row.of(1, "HiHi", "Hi6"),
+                                                Row.of(1, "HiHi", "Hi8"),
+                                                Row.of(null, "HeHe", "Hi9"))
+                                        .consumedAfterRestore(Row.of(2, "PostRestoreRight", "Hi5"))
+                                        .build())
+                        .runSql(
+                                "insert into MySink "
+                                        + "SELECT t2.a, t2.c, t1.c\n"
+                                        + "FROM (\n"
+                                        + " SELECT if(a = 3, cast(null as int), a) as a, b, c FROM T1\n"
+                                        + ") as t1\n"
+                                        + "JOIN (\n"
+                                        + " SELECT if(a = 3, cast(null as int), a) as a, b, c FROM T2\n"
+                                        + ") as t2\n"
+                                        + "ON \n"
+                                        + "  ((t1.a is null AND t2.a is null) OR\n"
+                                        + "  (t1.a = t2.a))\n"
+                                        + "  AND t1.b > t2.b")
+                        .build();
+
+        JOIN =
+                TableTestProgram.of("join", "test join")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a3 varchar", "b4 varchar")
+                                        .consumedBeforeRestore(
+                                                Row.of("Hi", "Hallo"),
+                                                Row.of("Hello", "Hallo Welt"),
+                                                Row.of("Hello world", "Hallo Welt"),
+                                                Row.of("Hi", "Hallo Welt wie gehts"))
+                                        .consumedAfterRestore(
+                                                Row.of("Hello there", "Hallo Welt wie"))
+                                        .build())
+                        .runSql("insert into MySink " + "SELECT a3, b4 FROM A, B WHERE a2 = b2")
+                        .build();
+
+        INNER_JOIN =
```

Comment:
Can we use the same input/output as `JOIN`. Plans should be the same, no?

Reply:
Hmm...  the syntax and plans are slightly different.

`INNER_JOIN` outputs `a1` and `b1` and uses those columns for the join condition.  

`JOIN` outputs `a3` and `b4` and uses different columns for the join condition.  That difference adds a `calc` node. 

In my first pass, I was trying to copy tests 1-1.  If these tests/plans are "close enough", I'd suggest that we could delete `INNER_JOIN`.  (That would address the other comment about readability for this test.) 

What do you think?

Reply:
I thought the difference is about `A JOIN B ON ...` vs `A, B WHERE ....`. I thought we have those two tests so that we check:
1. explicit join
2. having the filter condition results in an inner join instead of a cross-join with a filter

If that's the case we don't really care about the calc node. In the end we test the `JOIN` node here.